Bmx.Idp.Okta/OktaClient.cs

```diff
@@ -46,37 +58,62 @@ public class OktaClient : IIdentityProvider {
 
 			var authResp =
 				await AuthenticateChallengeMfaOkta(
-					new AuthenticateChallengeMfaOptions( mfaFactor.Id, challengeResponse, stateToken ) );
+					new AuthenticateChallengeMfaOptions( mfaFactor.Id, challengeResponse, _oktaStateToken ) );
 
-			_authenticateResponses.Add( AuthenticationStatus.SUCCESS, authResp );
+			_oktaSessionToken = authResp.SessionToken;
 
 			// TODO: Check for auth successes and return state
 			return true;
-			// TODO: Implicitly Call Session here
+		}
+
+		public async Task<string[]> GetAccounts( string accountType ) {
+			// TODO: Use existing session if it exists in ~/.bmx and isn't expired
+			var sessionResp = await CreateSessionOkta( new SessionOptions( _oktaSessionToken ) );
+			AddSession( sessionResp.Id );
+
+			_oktaApps = await GetAccountsOkta( sessionResp.UserId );
+			return _oktaApps.Where( app => app.AppName == accountType ).Select( app => app.Label ).ToArray();
+		}
+
+		private void AddSession( string sessionId ) {
+			CookieContainer.Add( new Cookie( "sid", sessionId, "/", HttpClient.BaseAddress.Host ) );
 		}
 
 		// TODO: Consider consolidating this kind of thing, ex: to a OktaHttpClient
 		private async Task<AuthenticateResponseInital> AuthenticateOkta( AuthenticateOptions authOptions ) {
 			var resp = await HttpClient.PostAsync( "authn",
-				new StringContent( JsonSerializer.Serialize( authOptions, _serializeOptions ) ) );
+				new StringContent( JsonSerializer.Serialize( authOptions, _serializeOptions ), Encoding.Default,
+					"application/json" ) );
+
 			return await JsonSerializer.DeserializeAsync<AuthenticateResponseInital>(
 				await resp.Content.ReadAsStreamAsync(), _serializeOptions );
 		}
 
 		private async Task<AuthenticateResponseSuccess> AuthenticateChallengeMfaOkta(
 			AuthenticateChallengeMfaOptions authOptions ) {
 			var resp = await HttpClient.PostAsync( $"authn/factors/{authOptions.FactorId}/verify",
-				new StringContent( JsonSerializer.Serialize( authOptions, _serializeOptions ) ) );
+				new StringContent( JsonSerializer.Serialize( authOptions, _serializeOptions ), Encoding.Default,
+					"application/json" ) );
 			return await JsonSerializer.DeserializeAsync<AuthenticateResponseSuccess>(
 				await resp.Content.ReadAsStreamAsync(), _serializeOptions );
 		}
+
+		private async Task<OktaSession> CreateSessionOkta( SessionOptions sessionOptions ) {
+			var resp = await HttpClient.PostAsync( "sessions",
+				new StringContent( JsonSerializer.Serialize( sessionOptions, _serializeOptions ), Encoding.Default,
+					"application/json" ) );
+			return await JsonSerializer.DeserializeAsync<OktaSession>( await resp.Content.ReadAsStreamAsync(),
+				_serializeOptions );
+		}
+
+		private async Task<OktaApp[]> GetAccountsOkta( string userId ) {
+			var resp = await HttpClient.GetAsync( $"users/{userId}/appLinks" );
+			return await JsonSerializer.DeserializeAsync<OktaApp[]>( await resp.Content.ReadAsStreamAsync(),
+				_serializeOptions );
```

Comment:
Minor: In case you haven't considered this already: it looks like `HttpClientJsonExtensions` contains a `GetFromJson()` you can use. Not sure it supports the same serialization options though.

Reply:
It does, doesn't seem to be present in .NET Core 3.1, added in 5.0+.

Reply:
Ah gotcha. Sorry forgot to check version numbers of the docs.